source/slang/slang-ir-com-interface.cpp

```diff
@@ -41,19 +42,17 @@ struct ComInterfaceLoweringContext
         }
     }
 
-    IRComPtrType* processInterfaceType(IRInterfaceType* type)
+    IRType* processInterfaceType(IRInterfaceType* type)
     {
         if (!type->findDecoration<IRComInterfaceDecoration>())
             return nullptr;
-
-        IRComPtrType* result = nullptr;
-
-        if (comPtrTypes.TryGetValue(type, result))
-            return result;
-
+     
         IRBuilder builder(sharedBuilder);
         builder.setInsertInto(module->getModuleInst());
-        result = builder.getComPtrType(type);
+
+        IRType* result = (artifactStyle == ArtifactStyle::Kernel) ?
+            static_cast<IRType*>(builder.getPtrType(type)) :
+            static_cast<IRType*>(builder.getComPtrType(type));
```

Comment:
Help me make sure I understand this, because from what I'm seeing this looks worrying. It *seems* like this is saying taht for "kernel" code we are deciding that COM interface types map to ordinary "raw" pointers, while for other cases of they stay as reference-counted `ComPtr<T>`?

I'm not sure I understand the logic here. The semantics of a `[COM] interface` type should always be those of `ComPtr<T>`. Even in a CPU kernel, wouldn't we want reference-counting to ensure things don't get released prematurely?

I was honestly expecting to see a special carve-out just for uses of interface types in function parameter lists.

Reply:
Yeah, I feel like it needs to be an error if a COM object is referenced in kernel code.

Reply:
Thanks for the context. I have no problem with this change in the short term, and I just wanted to make sure I understood the motivation.

In the long run I think the following are true:

1. We need a Slang `IFoo` for a `[COM] interface IFoo` to be 100% binary compatible/equivalent to a C++ `IFoo*`, in all contexts (function parameter/return, data structures, etc.). There is no wiggle room on that.
2. As a result of (1), we are going to need to start emitting the reference-count increment/decrement operations directly into Slang IR (which allows us to perform the relevant optimizations on them...).

Once we have (2), the surface area that we have to include in the prelude for reference-counting support is a lot smaller. In the worst case we are just talking about a `static inline` function that wraps the call to `AddRef` or `Release`, but maybe we don't even need that (and could emit them directly).

Reply:
Sounds good. 

Although it may not be easy (see ARC) doing reference counting in IR I think will be a significant overall win. 
You can't implement a 'perfect' ComPtr in C++. It always leads to unnecessary counting/busy work, and requires extra user code to improve perf which can introduce bugs.